mailer/engine.py

```diff
@@ -112,6 +113,28 @@ def release_lock(lock):
     logging.debug("released.")
 
 
+def error_handler(connection, message, err):
+    if not isinstance(err, (
+            socket_error,
+            smtplib.SMTPSenderRefused,
+            smtplib.SMTPRecipientsRefused,
+            smtplib.SMTPDataError,
+            smtplib.SMTPAuthenticationError,
+    )):
+        # TODO: Avoid endless retrying sending of messages on unrecognized
+        #       types of errors.
+        pass
```

Comment:
This code at the moment does nothing (`pass` just means it will carry on to the line below). Did you mean `raise`? I think there might be a problem using a bare `raise` here because the code is in a different scope from where the exception was caught, so it might need to use `raise err`.

As written, this code catches all exceptions and makes the messages deferred. This changes the behaviour from before - the previous error handler only caught recognized errors, others propagated.

The previous behaviour had some good features that we don't want to break. In particular, in the case of transient network errors SMTPConnectError the previous behaviour was better.

But there are some changes we do want to make to the old behaviour.

I think this comment gets it right: 
* https://github.com/pinax/django-mailer/issues/73#issuecomment-265561162 .
With some caveats:
* My comment here: https://github.com/pinax/django-mailer/issues/73#issuecomment-312021658
* Instead of 'cleanly abort send_all', I'm happy with re-raising the original (transient) error, which will then cause the `send_all` call to be aborted and the `manage.py` command to exit as an error, which is actually what we want.

I think it should be possible to add tests for all this, checking that the default error handler does all the right things.